### PR TITLE
s4744245_2DUNET_PROJ

### PR DESCRIPTION
**Pull request by JRunch**
This pull request contains an implementation of a 2D Unet on the CSIRO HIPMRI study. My implementation has achieve the goal of the 0.75 dice score on each of the 6 classes. The implementation is all described in the README file and the pull request contains the following files all under the recognition/2DUnet_Prostate_s4744245 file path:
dataset.py
modules.py
train.py
predict.py
util.py

### CODE REVIEW

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: The problem is well solved, all labels achieved DICE score over 0.75. In Traing and Validation Loss figure, why did the validation loss increase while training loss decreased?
2. **Implementation**: The load_data function in dataset.py script has many hard-coded paths within the function, it is better to set a path variable. [Hpyerparameters](https://github.com/JRunch/PatternAnalysis-2024/blob/topic-recognition/recognition/2DUnet_Prostate_s4744245/util.py#L10-L14) were defined and hard-coded in util.py, it would be better to put them at training script.
3. **Design**: Good
4. **Comment**: Basic comments, some functions do not have docstring.
5. **Difficulty**: Easy

**Comment by JRunch** [comment]
Not sure if I am meant to respond but in regards to the question about the Training and Validation Loss figure, the validation loss increasing while the training loss decreases towards the end suggests that the model is overfitting to the training data. Although this is bad, the final model chosen is that with the best validation dice coefficient meaning that it should be the best model before significant overfitting. In my final training, the best validation dice coefficient occurred at epoch 26 which was my saved best model and training was stopped at epoch 36 as I had a stopping criteria if the validation dice coefficient did not improve for 10 epochs.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Some minor inconsistancies seen.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. </td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Easy. UNet2D (Easy Difficulty)</td><td>-10</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. </td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. References missing.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). Submitted on time.</td><td></td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Remove cache files for merge.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-17</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by hanemma7moud** [comment]
Jack Runchel

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally, feedback applied +2
